File: crawler/rocketpunch/rocketpunch.go
package rocketpunch

import (
	"context"
	"encoding/json"
	"geekermeter-data/crawler"
	"io/ioutil"
	"log"
	"regexp"
	"strconv"
	"time"

	"github.com/chromedp/cdproto/cdp"
	"github.com/chromedp/chromedp"
)

var (
	baseURL = "https://www.rocketpunch.com/jobs?page=1"
)

func Rocketpunch() []crawler.Job {
	now := strconv.Itoa(time.Now().Year())
	var crawledData []crawler.Job
	ctx, cancel := chromedp.NewContext(context.Background())

	defer cancel()

	var loc string
	err := chromedp.Run(ctx,
		chromedp.Location(&loc),
		chromedp.Navigate(baseURL),
	)
	crawler.ErrHandler(err)

	var totalPageNode []*cdp.Node
	var totalPage string

	err = chromedp.Run(ctx,
		chromedp.Nodes("#search-results > div.ui.blank.right.floated.segment > div > div.tablet.computer.large.screen.widescreen.only > a:nth-child(7)",
			&totalPageNode, chromedp.ByID))
	crawler.ErrHandler(err)
	for _, row := range totalPageNode {
		totalPage = row.AttributeValue("data-query-add")[5:]
	}
	t, _ := strconv.Atoi(totalPage)
	for i := 1; i <= t; i++ { //페이지 단위
		log.Println("Current Page : ", i)
		temp := make([]crawler.Job, 200)
		var nodes []*cdp.Node
		var detailNode []*cdp.Node
		var newbieNode []*cdp.Node
		var dateNode []*cdp.Node
		var origin string
		sliceCap := 0

		err := chromedp.Run(ctx, //해당 페이지의 company Node가
			chromedp.Navigate(`https://www.rocketpunch.com/jobs?page=`+strconv.Itoa(i)),
			chromedp.Sleep(2*time.Second),
			chromedp.Nodes("#company-list > div.company.item",
				&nodes, chromedp.ByQueryAll),
		)

		for _, row := range nodes { //한 기업에서의 url 수집
			nodeNum := crawler.ExtractNum(row.PartialXPathByID())
			err = chromedp.Run(ctx,
				chromedp.Nodes("#company-list > div:nth-child("+nodeNum+") > div.content > div.company-jobs-detail > div.job-detail > div > a.nowrap.job-title.primary.link",
					&detailNode),
				chromedp.Nodes("#company-list > div:nth-child("+nodeNum+") > div.content > div.company-jobs-detail > div.job-detail > div > span.job-stat-info",
					&newbieNode),
				chromedp.Nodes("#company-list > div:nth-child("+nodeNum+") > div.content > div.company-jobs-detail > div.job-detail > div.job-dates",
					&dateNode),
				chromedp.Text(`#company-list > div:nth-child(`+nodeNum+`) > div.content > div.company-name > a:nth-child(1) > h4 > strong`,
					&origin),
			)
			crawler.ErrHandler(err)

			tempSliceCap := sliceCap
			for _, row := range detailNode {
				temp[sliceCap].Title = row.Children[0].NodeValue
				temp[sliceCap].URL = "https://www.rocketpunch.com" + row.AttributeValue("href")
				temp[sliceCap].Origin = origin
				sliceCap++
			}
			sliceCap = tempSliceCap

			for _, row := range newbieNode {
				//int8 로 반환
				temp[sliceCap].Newbie = crawler.Getnewbie(row.Children[0].NodeValue)
				//crawler.Gentnewbie에서 목표값으로 반환
				//temp[sliceCap].Newbie = crawler.Newbie(crawler.Getnewbie(row.Children[0].NodeValue))
				log.Println(temp[sliceCap].Newbie)
				sliceCap++
			}
			sliceCap = tempSliceCap

			for _, row := range dateNode {
				//var tt string
				var kk []*cdp.Node
				p := strconv.Itoa(int(row.ChildNodeCount))
				_ = chromedp.Run(ctx,
					chromedp.Nodes(row.FullXPath()+"/span["+p+"]", &kk),
				)
				r, _ := regexp.Compile("등록")
				for _, pow := range kk {
					if r.MatchString(pow.Children[0].NodeValue) {
						temp[sliceCap].StartDate = now + crawler.ExceptKorean(pow.Children[0].NodeValue)
					}
				}
				sliceCap++
			}
		}
		temp = temp[0:sliceCap]
		crawledData = append(crawledData, temp...)
	}

	return crawledData
}

func BodyText(box crawler.Job, forname int) {
	ctx, cancel := chromedp.NewContext(context.Background())

	defer cancel()

	// run task list
	var loc string
	err := chromedp.Run(ctx,
		chromedp.Navigate(box.URL),
		chromedp.Location(&loc),
	)
	crawler.ErrHandler(err)
	log.Println(box.URL)
	var nodes []*cdp.Node
	var clickNodes []*cdp.Node
	err = chromedp.Run(ctx,
		chromedp.Nodes("#wrap > div.eight.wide.job-content.column > section > h4 > a:nth-child(1)",
			&nodes))
	crawler.ErrHandler(err)

	clickctx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()
	err = chromedp.Run(clickctx,
		chromedp.Nodes("#wrap > div.eight.wide.job-content.column > section> div > a.see-more-text",
			&clickNodes),
	)

	for _, v := range clickNodes {
		err = chromedp.Run(ctx,
			chromedp.Click(v.FullXPath()),
		)
		crawler.ErrHandler(err)
	}

	var want string
	accepted := []string{"주요 업무", "업무 관련 기술 / 활동 분야", "채용 상세"}
	box.Content = make([]string, 3)
	count := 0
	for _, v := range nodes {
		err = chromedp.Run(ctx,
			chromedp.Text(v.FullXPath(), &want),
		)
		_, found := Find(accepted, want)
		if found == true {
			err = chromedp.Run(ctx,
				chromedp.Text(v.Parent.Parent.FullXPath(), &box.Content[count]),
			)
			crawler.ErrHandler(err)
			count++
		}
	}

	box.Content = box.Content[:count]
	toJson, _ := json.Marshal(box)
	t := strconv.Itoa(forname)
	_ = ioutil.WriteFile("./dataset/tmp/"+t+".json", toJson, 0644)

}

func Find(slice []string, val string) (int, bool) {
	for i, item := range slice {
		if item == val {
			return i, true
		}
	}
	return -1, false
}

//datetime은 크롤링의 대상이되는 날짜로 들어온다.
func Start(forname int, datetime string) int {
	log.Println("Start crawl Rocketpunch")
	list := Rocketpunch()
	log.Println("End crawl Rocketpunch")
	for _, row := range list {
		if datetime == row.StartDate {
			BodyText(row, forname)
			forname++
		}
	}
	return forname
}

File: crawler/rocketpunch/rocketpunch.py
import re, os, sys
import selenium
sys.path.append(os.path.dirname(os.path.abspath(os.path.dirname(__file__))))
from util import *
from selenium import webdriver
from bs4 import BeautifulSoup
import warnings
warnings.filterwarnings(action='ignore')


def rocketpunch(driver_path=None):
    if not driver_path:
        driver_path='chromedriver'
    driver = headless(driver_path)
#    driver = webdriver.Chrome(driver_path)
    driver.get('https://www.rocketpunch.com/jobs?page=1')
    final_data = []
    while True:
        crawled_data = []
        WebDriverWait(driver, 10).until(
            EC.presence_of_element_located((By.ID, "company-list"))
        )
        html = driver.page_source
        soup = BeautifulSoup(html,'html.parser')

        company_list = soup.select('#company-list > div.company.item')
        backup_url = driver.current_url
        for company in company_list:
            company_name = company.select('div.content > div.company-name > a > h4 > strong')[0].text
            post_list = company.select('div.content > div.company-jobs-detail > div.job-detail')
            for post in post_list:
                post_date = transfrom_date(post.select('div.job-dates > span')[-1].text,True)
                if not post_date:
                    continue
                post_main = post.select('div > a.nowrap.job-title.primary.link')[0]
                post_title = post_main.text
                post_url = 'https://www.rocketpunch.com' + post_main.get('href')
                post_newbie = make_newbie(post.select('div > span.job-stat-info')[0].text.replace(',',' ').split())
                crawled_data.append(
                    job(post_title,post_url,company_name,post_date,post_newbie).data
                )
        final_data.extend(crawled_data)
        _next = driver.find_elements_by_css_selector('#search-results > div.ui.blank.right.floated.segment > div.ui.pagination.menu > a')[-1].get_attribute('href')
        if _next:
            driver.get(_next)
        else:
            break

    return final_data, driver

def body_text(driver,json):
    driver.get(json['url'])
    WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.CSS_SELECTOR, "#wrap > div.eight.wide.job-content.column > section.row > h4"))
    )
    html = driver.page_source
    soup = BeautifulSoup(html,'html.parser')
    for section in soup.select('#wrap > div.eight.wide.job-content.column > section.row > h4'):
        if section.text in set(["주요 업무", "업무 관련 기술 / 활동 분야", "채용 상세"]):
            json['contents'] += section.parent.text.replace('\n',' ')

    return json

def run(driver_path=None):

    print('start crawling : rocketpunch')
    json_list, driver = rocketpunch(driver_path)
    print('The number of rocketpunch post : %d'%(len(json_list)))
    for i, json in enumerate(json_list):
        if i%10 == 0:
            print('rocketpunch post : %d'%(i))
        json = body_text(driver,json)
    driver.quit()
    print('finish crawling : rocketpunch')

    return json_list


if __name__ == "__main__":
    run()


    #rocketpunch()
    #driver = headless('/Users/mingihong/chromedriver')
    #t = job(0,'https://www.rocketpunch.com/jobs/43074/Trading-Business-Development-at-Tridge    ',0,0,0)
    #t = body_text(driver,t)
    #print(t.data)
